### PR TITLE
[misc] Add an option to print statistics into files

### PR DESCRIPTION
**Pull request by xumingkuan**
I don't know the file name of which we're going to print, so I print the statistics to the screen for now.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/program/program.cpp [hunk 1]
```diff
@@ -476,6 +477,24 @@ Kernel &Program::get_snode_writer(SNode *snode) {
...
+    if (current_test != nullptr) {
+      std::string file_name = current_test;
+      auto last_colon_pos = file_name.find_last_of(':');
+      TI_ASSERT(last_colon_pos != file_name.npos);
+      file_name = file_name.substr(last_colon_pos + 1);
+      auto first_space_pos = file_name.find_first_of(' ');
+      TI_ASSERT(first_space_pos != file_name.npos);
+      file_name = file_name.substr(0, first_space_pos);
+      file_name += ".log";
+      std::ofstream ofs(file_name);
```
Comment (xumingkuan) [inline taichi/program/program.cpp:491]: What if there are two tests with the same name in different files?

**Comment by xumingkuan** [comment]
I found that there are no kernels in `test_ad_basics.py::test_size1`, and `~Program()` won't be called in that test, so the number of the `.log` files is less than the number of the tests.

File: taichi/program/program.cpp (hunk 1 above)
Comment (xumingkuan) [reply taichi/program/program.cpp]: Now I also use the name of the test file.

**Comment by xumingkuan** [comment]
With `advanced_optimization`:
```
codegen_offloaded_tasks: 1644.00
codegen_statements  : 42375.00
```
Without  `advanced_optimization`:
```
codegen_offloaded_tasks: 1646.00
codegen_statements  : 45600.00
```

File: taichi/program/program.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/program/program.cpp]: Sounds good.

**Comment by yuanming-hu** [comment]
> With `advanced_optimization`:
> 
> ```
> codegen_offloaded_tasks: 1644.00
> codegen_statements  : 42375.00
> ```
> 
> Without `advanced_optimization`:
> 
> ```
> codegen_offloaded_tasks: 1646.00
> codegen_statements  : 45600.00
> ```

Cool! I would suggest list the ratio of instructions with/without advanced optimization, and we will have a feeling on 
 - Which test cases are significantly improved with advanced optimizations?
 - Which test cases are not improved at all? Is there any chances in those tests?

**Pull request merged**